src/node.cc

```diff
@@ -1066,17 +995,24 @@ MaybeLocal<Value> InternalMakeCallback(Environment* env,
     return Undefined(env->isolate());
   }
 
+  Local<Function> domain_cb = env->domain_callback();
   MaybeLocal<Value> ret;
-
-  {
+  if (asyncContext.async_id != 0 || domain_cb.IsEmpty() || recv.IsEmpty()) {
     ret = callback->Call(env->context(), recv, argc, argv);
-
-    if (ret.IsEmpty()) {
-      // NOTE: For backwards compatibility with public API we return Undefined()
-      // if the top level call threw.
-      scope.MarkAsFailed();
-      return scope.IsInnerMakeCallback() ? ret : Undefined(env->isolate());
+  } else {
+    Local<Array> argv_array = Array::New(env->isolate(), argc);
+    for (int i = 0; i < argc; i++) {
+      argv_array->Set(env->context(), i, argv[i]).FromJust();
     }
+    Local<Value> domain_argv[] = { callback, argv_array };
+    ret = domain_cb->Call(env->context(), recv, 2, domain_argv);
```

Comment:
It's probably faster to use spread in JS land and skip allocating the JS array here.  I.e.:
```c++
std::vector<Local<Value>> args(1 + argc);
args[0] = callback;
std::copy(&argv[0], &argv[argc], &args[1]);
ret = domain_cb->Call(env->context(), recv, args.size(), &args[0]);
```
Then in JS:
```js
function topLevelDomainCallback(cb, ...args) {
  // ...
}
```